File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.cpp
class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        int n=nums.size();
        vector<int> pre(n);
        for(int i=n-1;i>=0;i--){
            if(i==n-1) pre[i]=nums[i];
            else{
                pre[i]=pre[i+1]+nums[i];
            }
        }
        int count=0;
        unordered_map<int,int> mp;
        for(int i=n-1;i>=0;i--){
            if(pre[i]==k) count++;

            int rem=pre[i]-k;

            if(mp.find(rem)!=mp.end()) count+=mp[rem];

            mp[pre[i]]++;

            // cout<<i<<" "<<count<<endl;
        }

        return count;
    }
};

File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.java
class Solution {
    public int subarraySum(int[] nums, int k) {
        int n=nums.length;

        Map<Integer,Integer> mp=new HashMap<>();

        int[] pre=new int[n];
        pre[0]=nums[0];

        for(int i=1;i<n;i++){
            pre[i]=pre[i-1]+nums[i];
        }

        int cnt=0;
        for(int i=0;i<n;i++){
            if(pre[i]==k) cnt++;
            int rem=pre[i]-k;

            if(mp.containsKey(rem)){
                cnt+=mp.get(rem);
            }
            mp.put(pre[i],mp.getOrDefault(pre[i],0)+1);
        }
        return cnt;
    }
}